src/main/kotlin/org/jetbrains/research/ml/ast/transformations/outerNotElimination/NotEliminationRules.kt

```diff
@@ -0,0 +1,117 @@
+package org.jetbrains.research.ml.ast.transformations.outerNotElimination
+
+import com.jetbrains.python.PyTokenTypes
+import com.jetbrains.python.psi.*
+import org.jetbrains.research.ml.ast.transformations.PerformedCommandStorage
+import org.jetbrains.research.ml.ast.transformations.safePerformCommand
+
+private val PyExpression.isNegationExpression: Boolean
+    get() {
+        return this is PyPrefixExpression && operator == PyTokenTypes.NOT_KEYWORD
+    }
+
+private val PyPrefixExpression.innerOperand: PyExpression?
+    get() {
+        var inner = this.operand
+        while (inner is PyParenthesizedExpression)
+            inner = inner.containedExpression
+
+        return inner
+    }
+
+private fun PyElementGenerator.createNegationExpression(
+    level: LanguageLevel,
+    expression: PyExpression
+): PyPrefixExpression {
+    val newExpr = createExpressionFromText(level, "not x") as PyPrefixExpression
```

Comment:
It looks a little strange to me, it seems that you are always creating the same vertex with the text `not x,` but it is not. Could you please add some comment?

Reply:
In addition, I think you can look at the branch with the constant-folding transformation (it will be in the master branch in the coming days), since functions are already used there, for example, to place an expression in brackets. I think that we should reuse common functions in all places

Reply:
As I know, it is a standard way to create a PsiElement. Take a look at `PyElementGeneratorImpl.createBinaryExpression`,